File: src/jvmMain/java/org/angproj/crypt/ellipticcurve/Curve.java
/**
 * Copyright (c) 2019 Stark Bank S.A.
 *
 * This software is available under the terms of the MIT license.
 * The legal terms are attached to the LICENSE file and are made available on:
 *
 *      https://opensource.org/licenses/MIT
 *
 * SPDX-License-Identifier: MIT
 *
 * Contributors:
 *      Rafael Stark - original implementation
 *      Dalton Menezes - original implementation
 *      Caio Dottori - original implementation
 *      Thales Mello - original implementation
 *      Kristoffer Paulsson - adaption to Angelos Project
 */
package org.angproj.crypt.ellipticcurve;

import java.math.BigInteger;
import java.util.*;

import static org.angproj.crypt.ellipticcurve.CurveKt.*;

/**
 * Elliptic Curve Equation.
 * y^2 = x^3 + A*x + B (mod P)
 *
 */

public class Curve {

    public BigInteger A;
    public BigInteger B;
    public BigInteger P;
    public BigInteger N;
    public Point G;
    public String name;
    public long[] oid;
    public int digestSize;

    /**
     *
     * @param A A
     * @param B B
     * @param P P
     * @param N N
     * @param Gx Gx
     * @param Gy Gy
     * @param name name
     * @param oid oid
     */
    public Curve(BigInteger A, BigInteger B, BigInteger P, BigInteger N, BigInteger Gx, BigInteger Gy, String name, long[] oid, int digestSize) {
        this.A = A;
        this.B = B;
        this.P = P;
        this.N = N;
        this.G = new Point(Gx, Gy);
        this.name = name;
        this.oid = oid;
        this.digestSize = digestSize;
    }

    /**
     * Verify if the point `p` is on the curve
     *
     * @param p Point p = Point(x, y)
     * @return true if point is in the curve otherwise false
     */
    public boolean contains(Point p) {
        if (p.x.compareTo(BigInteger.ZERO) < 0) {
            return false;
        }
        if (p.x.compareTo(this.P) >= 0) {
            return false;
        }
        if (p.y.compareTo(BigInteger.ZERO) < 0) {
            return false;
        }
        if (p.y.compareTo(this.P) >= 0) {
            return false;
        }

        return p.y.pow(2).subtract(p.x.pow(3).add(A.multiply(p.x)).add(B)).mod(P).intValue() == 0;
    }

    /**
     *
     * @return int
     */
    public int length() {
        return (1 + N.toString(16).length()) / 2;
    }

    /**
     *
     */
    /*public static final Curve secp256k1 = new Curve(
        BigInteger.ZERO,
        BigInteger.valueOf(7),
        new BigInteger("fffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f", 16),
        new BigInteger("fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141", 16),
        new BigInteger("79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798", 16),
        new BigInteger("483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8", 16),
        "secp256k1",
        new long[]{1, 3, 132, 0, 10}
    );*/

    public static final Curve secp256k1 = secp256k1From();
    public static final Curve nistP192 = nistP192From();
    public static final Curve nistP224 = nistP224From();
    public static final Curve nistP256 = nistP256From();
    public static final Curve nistP384 = nistP384From();
    public static final Curve nistP521 = nistP521From();

    /**
     *
     */
    public static final List supportedCurves = new ArrayList();

    /**
     *
     */
    public static final Map curvesByOid = new HashMap();

    static {
        supportedCurves.add(secp256k1);
        supportedCurves.add(nistP192);
        supportedCurves.add(nistP224);
        supportedCurves.add(nistP256);
        supportedCurves.add(nistP384);
        supportedCurves.add(nistP521);

        for (Object c : supportedCurves) {
            Curve curve = (Curve) c;
            curvesByOid.put(Arrays.hashCode(curve.oid), curve);
        }
    }
}

File: src/jvmMain/kotlin/org/angproj/crypt/ellipticcurve/Curve.kt
/**
 * Copyright (c) 2024 by Kristoffer Paulsson <kristoffer.paulsson@talenten.se>.
 *
 * This software is available under the terms of the MIT license. Parts are licensed
 * under different terms if stated. The legal terms are attached to the LICENSE file
 * and are made available on:
 *
 *      https://opensource.org/licenses/MIT
 *
 * SPDX-License-Identifier: MIT
 *
 * Contributors:
 *      Kristoffer Paulsson - initial implementation
 */
package org.angproj.crypt.ellipticcurve

import org.angproj.crypt.sec.*
import org.angproj.aux.util.BinHex
import java.math.BigInteger

public fun secp256k1From(): Curve {

    val dp = Secp256Koblitz1.domainParameters
    return Curve(
        BigInteger(1, dp.a.toByteArray()),
        BigInteger(1, dp.b.toByteArray()),
        BigInteger(1, dp.p.toByteArray()),
        BigInteger(1, dp.n.toByteArray()),
        BigInteger(1, dp.G.x.toByteArray()),
        BigInteger(1, dp.G.y.toByteArray()),
        Secp256Koblitz1.name,
        longArrayOf(0, 1, 2, 3, 4),
        32
    )
}

/**
 * Bridging NIST curve implementation from Angelos Project to Starkbank implementation for testing and certifying.
 * */

public fun nistP192From(): Curve {
    val dp = Secp192Random1.domainParameters
    return Curve(
        BigInteger(1, dp.a.toByteArray()),
        BigInteger(1, dp.b.toByteArray()),
        BigInteger(1, dp.p.toByteArray()),
        BigInteger(1, dp.n.toByteArray()),
        BigInteger(1, dp.G.x.toByteArray()),
        BigInteger(1, dp.G.y.toByteArray()),
        Secp192Random1.name,
        longArrayOf(0, 1, 2, 3, 4),
        20
    )
}

public fun nistP224From(): Curve {
    val dp = Secp224Random1.domainParameters
    return Curve(
        BigInteger(1, dp.a.toByteArray()),
        BigInteger(1, dp.b.toByteArray()),
        BigInteger(1, dp.p.toByteArray()),
        BigInteger(1, dp.n.toByteArray()),
        BigInteger(1, dp.G.x.toByteArray()),
        BigInteger(1, dp.G.y.toByteArray()),
        Secp224Random1.name,
        longArrayOf(1, 1, 2, 3, 4),
        28
    )
}

public fun nistP256From(): Curve {
    val dp = Secp256Random1.domainParameters
    return Curve(
        BigInteger(1, dp.a.toByteArray()),
        BigInteger(1, dp.b.toByteArray()),
        BigInteger(1, dp.p.toByteArray()),
        BigInteger(1, dp.n.toByteArray()),
        BigInteger(1, dp.G.x.toByteArray()),
        BigInteger(1, dp.G.y.toByteArray()),
        Secp256Random1.name,
        longArrayOf(2, 1, 2, 3, 4),
        32
    )
}

public fun nistP384From(): Curve {
    val dp = Secp384Random1.domainParameters
    return Curve(
        BigInteger(1, dp.a.toByteArray()),
        BigInteger(1, dp.b.toByteArray()),
        BigInteger(1, dp.p.toByteArray()),
        BigInteger(1, dp.n.toByteArray()),
        BigInteger(1, dp.G.x.toByteArray()),
        BigInteger(1, dp.G.y.toByteArray()),
        Secp384Random1.name,
        longArrayOf(3, 1, 2, 3, 4),
        48
    )
}

public fun nistP521From(): Curve {
    val dp = Secp521Random1.domainParameters
    return Curve(
        BigInteger(1, dp.a.toByteArray()),
        BigInteger(1, dp.b.toByteArray()),
        BigInteger(1, dp.p.toByteArray()),
        BigInteger(1, dp.n.toByteArray()),
        BigInteger(1, dp.G.x.toByteArray()),
        BigInteger(1, dp.G.y.toByteArray()),
        Secp521Random1.name,
        longArrayOf(4, 1, 2, 3, 4),
        64
    )
}

public fun pointOnCurve2(c: Curve, p: Point): Boolean {
    if (p.x.compareTo(BigInteger.ZERO) < 0) {
        return false
    }
    if (p.x.compareTo(c.P) >= 0) {
        return false
    }
    if (p.y.compareTo(BigInteger.ZERO) < 0) {
        return false
    }
    if (p.y.compareTo(c.P) >= 0) {
        return false
    }

    println(BinHex.encodeToHex(p.x.toByteArray()))
    println(BinHex.encodeToHex(p.y.toByteArray()))


    val Amul: BigInteger = c.A.multiply(p.x)
    println(BinHex.encodeToHex(Amul.toByteArray()))

    val Xpow = p.x.pow(3)
    println(BinHex.encodeToHex(Xpow.toByteArray()))

    val Add1 = Xpow.add(Amul)
    println(BinHex.encodeToHex(Add1.toByteArray()))

    val Add2 = Add1.add(c.B)
    println(BinHex.encodeToHex(Add2.toByteArray()))


    val Ypow = p.y.pow(2)
    println(BinHex.encodeToHex(Ypow.toByteArray()))

    val Sub1 = Ypow.subtract(Add2)
    println(BinHex.encodeToHex(Sub1.toByteArray()))

    val modP = Sub1.mod(c.P)
    println(BinHex.encodeToHex(modP.toByteArray()))
    println("")

    return modP.toInt() == 0

    //return p.y.pow(2).subtract(p.x.pow(3).add(A.multiply(p.x)).add(B)).mod(P).intValue() == 0;
}